python/hail/linalg/matrix.py

```diff
@@ -47,27 +53,61 @@ def write(self, path):
     @handle_py4j
     @typecheck_method(that=block_matrix_type)
     def __mul__(self, that):
-        return BlockMatrix(self.hc,
-            self._jbm.multiply(that._jbm))
+        return BlockMatrix(self._jbm.multiply(that._jbm))
 
     @handle_py4j
     @typecheck_method(cols_to_keep=listof(integral))
     def filter_cols(self, cols_to_keep):
-        return BlockMatrix(self.hc,
-            self._jbm.filterCols(jarray(Env.jvm().long, cols_to_keep)))
+        return BlockMatrix(self._jbm.filterCols(jarray(Env.jvm().long, cols_to_keep)))
 
     @handle_py4j
     @typecheck_method(rows_to_keep=listof(integral))
     def filter_rows(self, rows_to_keep):
-        return BlockMatrix(self.hc,
-            self._jbm.filterRows(jarray(Env.jvm().long, rows_to_keep)))
+        return BlockMatrix(self._jbm.filterRows(jarray(Env.jvm().long, rows_to_keep)))
     
     @handle_py4j
     @typecheck_method(rows_to_keep=listof(integral),
                       cols_to_keep=listof(integral))
     def filter(self, rows_to_keep, cols_to_keep):
-        return BlockMatrix(self.hc,
-            self._jbm.filter(jarray(Env.jvm().long, rows_to_keep),
+        return BlockMatrix(self._jbm.filter(jarray(Env.jvm().long, rows_to_keep),
                              jarray(Env.jvm().long, cols_to_keep)))
 
+    @handle_py4j
+    def transpose(self):
+        return BlockMatrix(self._jbm.transpose())
+
+    def cache(self):
+        return BlockMatrix(self._jbm.cache())
+
+    @typecheck_method(storage_level=enumeration('NONE', 'DISK_ONLY', 'DISK_ONLY_2', 'MEMORY_ONLY',
+                                                'MEMORY_ONLY_2', 'MEMORY_ONLY_SER', 'MEMORY_ONLY_SER_2',
+                                                'MEMORY_AND_DISK', 'MEMORY_AND_DISK_2', 'MEMORY_AND_DISK_SER',
+                                                'MEMORY_AND_DISK_SER_2', 'OFF_HEAP'))
+    def persist(self, storage_level='MEMORY_AND_DISK'):
+        return BlockMatrix(self._jbm.persist(storage_level))
+
+    def unpersist(self):
+        return BlockMatrix(self._jbm.unpersist())
+
 block_matrix_type.set(BlockMatrix)
+
+
+@handle_py4j
+@typecheck(entry_expr=expr_numeric,
+           path=strlike,
+           block_size=nullable(integral))
+def block_matrix_from_expr(entry_expr,
```

Comment:
does this do what I think it does? Constructs a numeric block matrix from a numeric entry of a matrix? AWESOME!

Reply:
needs docs.

Reply:
do we _have_ to write/read for performance reasons, or is it possible to construct on the fly?

Reply:
The write/read is for shuffle-related reasons.

Reply:
Should this be `BlockMatrix.from_expr`?  (Like VariantDataset.from_table in api1).  Or from_matrix_table_expr?  Why isn't the path a temporary?

Reply:
On BlockMatrix.from_expr:
I tried this first, but I found some problems with the fact that I'm using other BlockMatrix `@staticmethod`s such as `default_block_size` and `read`. I looked on the internet to solve this and basically what I found was a bunch of people saying "well that's not the pythonic way to do things" but not actually a good solution. Maybe I can solve this by making them classmethods instead? I'll experiment.
On path:
I kind of went back and forth on this and eventually decided "well, maybe someone will want to convert a VDS write this block matrix to a disk anyways and it would be nice to not make them copy it if they want a persistent version." I also wasn't super sure where to put this method, but there could conceivably be `MatrixTable.write_block_matrix(expr, path)` and this could just be a wrapper for that + BlockMatrix.read(path) with a temporary path. What do you think?

Reply:
I fixed the first one (and renamed `from_matrix_table`). Jon and I talked about the second one. I've changed the path option to be optional for now and default is to generate a temp file.